### ISSUE TITLE
dynamic configuration block from null resource triggers

### ISSUE THREAD

**Issue by theTranqu** [hashicorp/terraform#20692]
### Terraform Version

```
0.12.0-alpha4
```

### Terraform Configuration Files

```hcl

resource "aws_eip" "nlb" {
  count = (length(var.nlb_elastic_ip_allocation_id)==0 && var.lb_type=="nlb") ? var.amount_azs : 0
  vpc   = true
}


resource "aws_lb" "nlb" {
  name               = var.application_name
  count              = var.lb_type=="nlb" ? 1 : 0
  internal           = var.lb_internal
  load_balancer_type = "network"

  //generate subnet mapping making use of dynamic block
  dynamic "subnet_mapping" {
    for_each = [ for mapping in null_resource.subnet_mapping: {
      subnet_id = mapping.triggers.subnet_id
      allocation_id = mapping.triggers.allocation_id
    }]

    content {
      subnet_id = subnet_mapping.subnet_id
      allocation_id = subnet_mapping.allocation_id
    }
  }
  enable_cross_zone_load_balancing = var.nlb_enable_cross_zone_load_balancing
  idle_timeout                     = var.idle_timeout

  tags = local.tags
}

resource "null_resource" "subnet_mapping" {
  // calculate count. incase NLB is selected, evaluate: if given static elastic ips == 0 use var.amount_azs, otherwise amount of given ips
  count = var.lb_type=="nlb" ? (length(var.nlb_elastic_ip_allocation_id)==0 ? var.amount_azs : length(var.nlb_elastic_ip_allocation_id)): 0

  triggers = {
    //select either private or pub subnets, depending on lb_internal variable
    subnet_id = element(local.subnets_choice, count.index)

    //select EIP, either from given static EIPs or use dynamically created ips
    allocation_id = element(local.eip_choice, count.index)
  }
}


locals {
  subnets_choice = split(",",(var.lb_internal ? local.private_subnets : local.public_subnets))
  eip_choice     = length(var.nlb_elastic_ip_allocation_id) == 0 ? aws_eip.nlb.*.allocation_id : var.nlb_elastic_ip_allocation_id
}

locals {
  public_subnets = join(",",data.terraform_remote_state.network.outputs.subnets-public.subnet_id)
  

  private_subnets = join(",",data.terraform_remote_state.network.outputs.subnets-private.subnet_id)
}
```

### Expected Behavior

The null_resource should create a list of maps in its triggers attribute. This list of maps should definetly be iterable

### Actual Behavior

Error: Invalid dynamic for_each value
 
   on .terraform/modules/application/modules/aws/application/alb_nlb.tf line 15, in resource "aws_lb" "nlb":
     for_each = [ for mapping in null_resource.subnet_mapping: {
      subnet_id = mapping.triggers.subnet_id
      allocation_id = mapping.triggers.allocation_id
     }]
 
 Cannot use a value of type {{{}}} in for_each. An iterable collection is
 required.

### Additional Context

This happens inside of a module, that creates different types of loadbalancers, depending on a choice variable. Depending on whether the loadbalancer is supposed to be public or not, it is placed in different subnets. The module-user can either supply a list of elastic-ip allocation-ids or not. In the latter case, elastic ips should be created by the module and used in the NLB's subnet-mapping block, depending on the amount of availability-zones that should be used.

Is my assumption, that the triggers attribute of a null_resource with set count attribute should be an iterable list of maps wrong, or is there actually a bug here?

**Comment by apparentlymart** [comment]
Hi @theTranqu! Thanks for reporting this.

Would you mind trying this again with [the beta1 release](https://www.hashicorp.com/blog/announcing-terraform-0-1-2-beta1)? This issue seems familiar and I think it was already fixed since the alpha4 release. If not, it should at least give a better error message that doesn't include that strange placeholder for the type.

**Comment by theTranqu** [comment]
@apparentlymart Thank you very much for your assistance, using the beta-v1 version has indeed solved the problem that the list was not recognized as iterable.

```hcl

resource "aws_lb" "nlb" {
  name               = var.application_name
  count              = var.lb_type=="nlb" ? 1 : 0
  internal           = var.lb_internal
  load_balancer_type = "network"

  dynamic "subnet_mapping" {
    for_each = null_resource.subnet_mapping[*].triggers

    content {
      subnet_id = subnet_mapping.value.subnet_id
      allocation_id = subnet_mapping.value.allocation_id
    }
  }
  enable_cross_zone_load_balancing = var.nlb_enable_cross_zone_load_balancing
  idle_timeout                     = var.idle_timeout

  tags = local.tags
}
```
This does not break the terraform plan anymore.

Now I got some new warnings, because we are using one tfvars-file per vpc/environment, storing all the variables used for all possible jobs that we have....but I just now found out, that in the future TFVARS files can _only_ contain variables actually declared in the respective terraform job/folder - 
and I am stunned. This completely breaks our pipeline and we had no warning, that this might happen.
It was great to keep tfvars files in central locations and have all variables for a specific environment in one file. whelp, I guess it's time to split things up, isn't it?

Thank you @apparentlymart for your hint to try the new tf beta build. I guess this can be closed
